File: app/src/main/java/com/davismiyashiro/expenses/view/addexpense/ExpenseActivity.java
package com.davismiyashiro.expenses.view.addexpense;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.ActionBar;
import android.view.View;
import android.view.WindowManager;
import android.widget.EditText;

import com.davismiyashiro.expenses.Injection;
import com.davismiyashiro.expenses.datatypes.Tab;
import com.davismiyashiro.expenses.model.BaseCompatActivity;
import com.davismiyashiro.expenses.R;
import com.davismiyashiro.expenses.datatypes.Expense;
import com.davismiyashiro.expenses.view.splitter.SplitterActivity;

import timber.log.Timber;

public class ExpenseActivity extends BaseCompatActivity implements ExpenseView.View, View.OnClickListener {

    private static String TAB_REQUESTED = "com.davismiyashiro.expenses.view.expense";
    private static String EDIT_EXPENSE = "com.davismiyashiro.expenses.view.expense.edit";

    private TextInputLayout mInputLayoutDesc;
    private EditText mEditTextDescription;
    private TextInputLayout mInputLayoutValue;
    private EditText mEditTextValue;
    private FloatingActionButton mFabExpense;

    private ExpenseView.UserActionsListener mPresenter;

    private Tab mTab;
    private Expense mExpense;

    public static Intent newIntent (Context context, Tab tab, Expense expense) {
        Intent intent = new Intent(context, ExpenseActivity.class);
        intent.putExtra(TAB_REQUESTED, tab);
        intent.putExtra(EDIT_EXPENSE, expense);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setContentView(R.layout.expenses);
        super.onCreate(savedInstanceState);

        mInputLayoutDesc = (TextInputLayout) findViewById(R.id.edt_exp_desc_layout);
        mEditTextDescription = (EditText) findViewById(R.id.edtDescriptionName);
        requestFocus(mEditTextDescription);

        mInputLayoutValue = (TextInputLayout) findViewById(R.id.edt_exp_value_layout);
        mEditTextValue = (EditText) findViewById(R.id.edtValue);

        mFabExpense = (FloatingActionButton) findViewById(R.id.fab_add_expense_details);
        mFabExpense.setOnClickListener(this);
        mFabExpense.show();
//        ObjectAnimator fadeAnimX = ObjectAnimator.ofFloat(mFabExpense, View.SCALE_X, 0, 1);
//        fadeAnimX.setDuration(500);
//        //fadeAnimX.start();
//
//        ObjectAnimator fadeAnimY = ObjectAnimator.ofFloat(mFabExpense, View.SCALE_Y, 0, 1);
//        fadeAnimY.setDuration(500);
//        //fadeAnimY.start();
//
//        AnimatorSet set1 = new AnimatorSet();
//        set1.playTogether(fadeAnimX, fadeAnimY);
//        set1.start();

        mPresenter = new ExpensePresenterImpl(this, Injection.provideTabsRepository(this));

        mTab = getIntent().getParcelableExtra(TAB_REQUESTED);
        mExpense = getIntent().getParcelableExtra(EDIT_EXPENSE);

        ActionBar ab = getSupportActionBar();
        if (isExpenseNew()) {
            ab.setTitle(R.string.add_expense_action_bar);
        } else {
            ab.setTitle(R.string.edit_expense_action_bar);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        Timber.d("onResume");

        if (!isExpenseNew()){
            setExpenseDescription(mExpense.getDescription());
            setExpenseValue(String.valueOf(mExpense.getValue()));
        }
    }

    @Override
    public void chooseParticipants(Expense expense) {
        Intent intent = SplitterActivity.newIntent(this, expense);
        startActivity(intent);
    }

    @Override
    public void setExpenseDescription(String desc) {
        mEditTextDescription.setText(desc);
    }

    @Override
    public void setExpenseValue(String value) {
        mEditTextValue.setText(value);
    }

    @Override
    public void setDescriptionError(boolean value) {
        //mEditTextDescription.setError(getString(R.string.error_field_required));
        if (value) {
            mInputLayoutDesc.setError(getString(R.string.error_field_required));
            requestFocus(mEditTextDescription);
        }
        mInputLayoutDesc.setErrorEnabled(value);
    }

    @Override
    public void setValueError(boolean value) {
        //mEditTextValue.setError(getString(R.string.error_field_required));
        if (value) {
            mInputLayoutValue.setError(getString(R.string.error_field_required));
            requestFocus(mEditTextValue);
        }
        mInputLayoutValue.setErrorEnabled(value);
    }

    @Override
    public void onSuccess() {
        String description = mEditTextDescription.getText().toString();
        String valueStr = mEditTextValue.getText().toString();
        double value = Double.parseDouble(valueStr);
        Expense expense;

        if (isExpenseNew()) {

            expense = new Expense(description, value, mTab.getGroupId());

            mPresenter.addExpense(expense);
        } else {
            expense = Expense.retrieveExpense(mExpense.getId(), description, value, mTab.getGroupId());

            mPresenter.updateExpense(expense);
        }
        chooseParticipants(expense);

        // ---closes the activity---
        mFabExpense.hide();
        finish();
    }

    @Override
    public void onClick(View v) {
        mPresenter.validateExpenseFields(mEditTextDescription.getText().toString(), mEditTextValue.getText().toString());
    }

    private void requestFocus(View view) {
        if (view.requestFocus()) {
            getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        }
    }

    private boolean isExpenseNew () {
        return mExpense == null;
    }
}

File: app/src/main/java/com/davismiyashiro/expenses/view/addexpense/ExpenseActivity.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Davis Miyashiro
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.davismiyashiro.expenses.view.addexpense

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.TextInputLayout
import android.view.View
import android.view.WindowManager
import android.widget.EditText

import com.davismiyashiro.expenses.Injection
import com.davismiyashiro.expenses.datatypes.Tab
import com.davismiyashiro.expenses.view.BaseCompatActivity
import com.davismiyashiro.expenses.R
import com.davismiyashiro.expenses.datatypes.Expense
import com.davismiyashiro.expenses.view.splitter.SplitterActivity

import timber.log.Timber

class ExpenseActivity : BaseCompatActivity(), ExpenseView.View, View.OnClickListener {

    private lateinit var mInputLayoutDesc: TextInputLayout
    private lateinit var mEditTextDescription: EditText
    private lateinit var mInputLayoutValue: TextInputLayout
    private lateinit var mEditTextValue: EditText
    private lateinit var mFabExpense: FloatingActionButton

    private var mPresenter: ExpenseView.UserActionsListener? = null

    private var mTab: Tab? = null
    private var mExpense: Expense? = null

    private val isExpenseNew: Boolean
        get() = mExpense == null

    override fun onCreate(savedInstanceState: Bundle?) {
        setContentView(R.layout.expenses)
        super.onCreate(savedInstanceState)

        mInputLayoutDesc = findViewById(R.id.edt_exp_desc_layout)
        mEditTextDescription = findViewById(R.id.edtDescriptionName)
        requestFocus(mEditTextDescription)

        mInputLayoutValue = findViewById(R.id.edt_exp_value_layout)
        mEditTextValue = findViewById(R.id.edtValue)

        mFabExpense = findViewById(R.id.fab_add_expense_details)
        mFabExpense.setOnClickListener(this)
        mFabExpense.show()

        mPresenter = ExpensePresenterImpl(this, Injection.provideTabsRepository(this))

        mTab = intent.getParcelableExtra(TAB_REQUESTED)
        mExpense = intent.getParcelableExtra(EDIT_EXPENSE)

        val ab = supportActionBar
        if (isExpenseNew) {
            ab?.setTitle(R.string.add_expense_action_bar)
        } else {
            ab?.setTitle(R.string.edit_expense_action_bar)
        }
    }

    override fun onResume() {
        super.onResume()
        Timber.d("onResume")

        if (!isExpenseNew) {
            setExpenseDescription(mExpense?.description.toString())
            setExpenseValue(mExpense?.value.toString())
        }
    }

    override fun chooseParticipants(expense: Expense) {
        val intent = SplitterActivity.newIntent(this, expense)
        startActivity(intent)
    }

    override fun setExpenseDescription(desc: String) {
        mEditTextDescription.setText(desc)
    }

    override fun setExpenseValue(value: String) {
        mEditTextValue.setText(value)
    }

    override fun setDescriptionError(value: Boolean) {
        if (value) {
            mInputLayoutDesc.error = getString(R.string.error_field_required)
            requestFocus(mEditTextDescription)
        }
        mInputLayoutDesc.isErrorEnabled = value
    }

    override fun setValueError(value: Boolean) {
        if (value) {
            mInputLayoutValue.error = getString(R.string.error_field_required)
            requestFocus(mEditTextValue)
        }
        mInputLayoutValue.isErrorEnabled = value
    }

    override fun onSuccess() {
        val description = mEditTextDescription.text.toString()
        val valueStr = mEditTextValue.text.toString()
        val value = java.lang.Double.parseDouble(valueStr)
        val expense: Expense

        if (isExpenseNew) {

            expense = Expense(description, value, mTab?.groupId.toString())

            mPresenter?.addExpense(expense)
        } else {
            expense = Expense.retrieveExpense(mExpense?.id.toString(), description, value, mTab?.groupId.toString())

            mPresenter?.updateExpense(expense)
        }
        chooseParticipants(expense)

        // ---closes the activity---
        mFabExpense.hide()
        finish()
    }

    override fun onClick(v: View) {
        mPresenter?.validateExpenseFields(mEditTextDescription.text.toString(), mEditTextValue.text.toString())
    }

    private fun requestFocus(view: View) {
        if (view.requestFocus()) {
            window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
        }
    }

    companion object {

        private const val TAB_REQUESTED = "com.davismiyashiro.expenses.view.expense"
        private const val EDIT_EXPENSE = "com.davismiyashiro.expenses.view.expense.edit"

        fun newIntent(context: Context, tab: Tab, expense: Expense?): Intent {
            val intent = Intent(context, ExpenseActivity::class.java)
            intent.putExtra(TAB_REQUESTED, tab)
            intent.putExtra(EDIT_EXPENSE, expense)
            return intent
        }
    }
}
